File: frontend/src/lib/api/notes.ts
import type {
  Evaluation,
  GrilleNotes,
  MoyennesClasse,
  TypeEvaluation,
} from "@/types/notes";
import { apiFetch } from "./client";

export async function listTypesEvaluation(token: string) {
  return apiFetch<TypeEvaluation[]>("/api/v1/notes/types-evaluation", {}, token);
}

export async function listEvaluations(
  token: string,
  params: { classe_id: string; matiere_id: string; periode_id: string },
) {
  const q = new URLSearchParams(params);
  return apiFetch<Evaluation[]>(`/api/v1/notes/evaluations?${q}`, {}, token);
}

export async function createEvaluation(
  token: string,
  data: {
    libelle: string;
    classe_id: string;
    matiere_id: string;
    periode_id: string;
    type_evaluation_id: string;
    coefficient?: number;
  },
) {
  return apiFetch<Evaluation>("/api/v1/notes/evaluations", {
    method: "POST",
    body: JSON.stringify(data),
  }, token);
}

export async function getGrilleNotes(token: string, evaluationId: string) {
  return apiFetch<GrilleNotes>(`/api/v1/notes/evaluations/${evaluationId}/grille`, {}, token);
}

export async function saveNotes(
  token: string,
  evaluationId: string,
  notes: { eleve_id: string; valeur?: number | null; is_absent?: boolean; appreciation_libre?: string }[],
) {
  return apiFetch<GrilleNotes>(`/api/v1/notes/evaluations/${evaluationId}/notes`, {
    method: "PUT",
    body: JSON.stringify({ notes }),
  }, token);
}

export async function getMoyennesClasse(token: string, classeId: string, periodeId: string) {
  return apiFetch<MoyennesClasse>(
    `/api/v1/notes/classes/${classeId}/periodes/${periodeId}/moyennes`,
    {},
    token,
  );
}

export async function validerNotes(token: string, classeId: string, periodeId: string) {
  return apiFetch<{ verrouille: boolean }>(
    `/api/v1/notes/classes/${classeId}/periodes/${periodeId}/valider`,
    { method: "POST" },
    token,
  );
}

File: backend/app/api/v1/endpoints/notes.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, require_any_permission, require_permission
from app.models.user import User
from app.schemas.notes import (
    EvaluationCreate,
    EvaluationResponse,
    GrilleNotesResponse,
    MoyennesClasseResponse,
    NotesBulkUpdate,
    TypeEvaluationCreate,
    TypeEvaluationResponse,
    ValidationResponse,
)
from app.services import notes_service

router = APIRouter()

READ_PERMISSIONS = ("grades.modify", "grades.validate_bulletins", "reports.view_pedagogical")


@router.get("/types-evaluation", response_model=list[TypeEvaluationResponse])
async def list_types(
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await notes_service.list_types_evaluation(db)


@router.post("/types-evaluation", response_model=TypeEvaluationResponse, status_code=status.HTTP_201_CREATED)
async def create_type(
    data: TypeEvaluationCreate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_permission("grades.validate_bulletins")),
):
    return await notes_service.create_type_evaluation(db, data)


@router.get("/evaluations", response_model=list[EvaluationResponse])
async def list_evaluations(
    classe_id: UUID = Query(...),
    matiere_id: UUID = Query(...),
    periode_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await notes_service.list_evaluations(db, classe_id, matiere_id, periode_id)


@router.post("/evaluations", response_model=EvaluationResponse, status_code=status.HTTP_201_CREATED)
async def create_evaluation(
    data: EvaluationCreate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_permission("grades.modify")),
):
    return await notes_service.create_evaluation(db, data)


@router.get("/evaluations/{evaluation_id}/grille", response_model=GrilleNotesResponse)
async def get_grille(
    evaluation_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await notes_service.get_grille_notes(db, evaluation_id)


@router.put("/evaluations/{evaluation_id}/notes", response_model=GrilleNotesResponse)
async def update_notes(
    evaluation_id: UUID,
    data: NotesBulkUpdate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_permission("grades.modify")),
):
    return await notes_service.bulk_update_notes(
        db,
        evaluation_id,
        data,
        user_id=user.id,
        ip_address=request.client.host if request.client else None,
        user_email=user.email,
    )


@router.get("/classes/{classe_id}/periodes/{periode_id}/moyennes", response_model=MoyennesClasseResponse)
async def moyennes_classe(
    classe_id: UUID,
    periode_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await notes_service.get_moyennes_classe(db, classe_id, periode_id)


@router.post("/classes/{classe_id}/periodes/{periode_id}/valider", response_model=ValidationResponse)
async def valider_notes(
    classe_id: UUID,
    periode_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("grades.validate_bulletins")),
):
    return await notes_service.valider_periode(db, classe_id, periode_id, current_user.id)


@router.post("/classes/{classe_id}/periodes/{periode_id}/deverrouiller", response_model=ValidationResponse)
async def deverrouiller_notes(
    classe_id: UUID,
    periode_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("grades.validate_bulletins")),
):
    return await notes_service.valider_periode(
        db, classe_id, periode_id, current_user.id, verrouiller=False
    )
